From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 1 Mar 2025 14:14:51 +0100
Subject: [PATCH 047/225] power: supply: sc27xx: automatically calibrate on
 every boot
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The last saved capacity value may become outdated if the device is
powered off for a long time or booted in a state in which the fuel
gauge driver is not active. Instead of recomputing the capacity only
after a full shutdown, which happens very rarely on devices with a
non-removable battery, adjust it on every boot. Try to use a value
close to the saved one to avoid large jumps due to measurement
inaccuracies.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/power/supply/sc27xx_fuel_gauge.c | 42 +++++++++++++++---------
 1 file changed, 26 insertions(+), 16 deletions(-)

diff --git a/drivers/power/supply/sc27xx_fuel_gauge.c b/drivers/power/supply/sc27xx_fuel_gauge.c
index e7a2d43ce..023d7d701 100644
--- a/drivers/power/supply/sc27xx_fuel_gauge.c
+++ b/drivers/power/supply/sc27xx_fuel_gauge.c
@@ -295,22 +295,9 @@ static int sc27xx_fgu_read_last_cap(struct sc27xx_fgu_data *data, int *cap)
  */
 static int sc27xx_fgu_get_boot_capacity(struct sc27xx_fgu_data *data, int *cap)
 {
-	int volt, cur, oci, ocv, ret;
+	int volt, cur, oci, ocv, ocv_cap, ret;
 	bool is_first_poweron = sc27xx_fgu_is_first_poweron(data);
 
-	/*
-	 * If system is not the first power on, we should use the last saved
-	 * battery capacity as the initial battery capacity. Otherwise we should
-	 * re-calculate the initial battery capacity.
-	 */
-	if (!is_first_poweron) {
-		ret = sc27xx_fgu_read_last_cap(data, cap);
-		if (ret)
-			return ret;
-
-		return sc27xx_fgu_save_boot_mode(data, SC27XX_FGU_NORMAIL_POWERTON);
-	}
-
 	/*
 	 * After system booting on, the SC27XX_FGU_CLBCNT_QMAXL register saved
 	 * the first sampled open circuit current.
@@ -340,8 +327,31 @@ static int sc27xx_fgu_get_boot_capacity(struct sc27xx_fgu_data *data, int *cap)
 	 * Parse the capacity table to look up the correct capacity percent
 	 * according to current battery's corresponding OCV values.
 	 */
-	*cap = power_supply_ocv2cap_simple(data->cap_table, data->table_len,
-					   ocv);
+	ocv_cap = power_supply_ocv2cap_simple(data->cap_table, data->table_len,
+					      ocv);
+
+	/*
+	 * If system is not the first power on, we should use the last saved
+	 * battery capacity as the initial battery capacity. Otherwise we should
+	 * use the capacity value calculated above.
+	 */
+	if (!is_first_poweron) {
+		ret = sc27xx_fgu_read_last_cap(data, cap);
+		if (ret)
+			return ret;
+
+		/*
+		 * Limit the saved capacity to a reasonable range based on
+		 * the calculated estimate. This is needed because the saved
+		 * value may be outdated.
+		 */
+		if (*cap < ocv_cap - 5)
+			*cap = ocv_cap - 5;
+		else if (*cap > ocv_cap + 5)
+			*cap = ocv_cap + 5;
+	} else {
+		*cap = ocv_cap;
+	}
 
 	ret = sc27xx_fgu_save_last_cap(data, *cap);
 	if (ret)
